### PR TITLE
Fix BigNumber issues

### PR DESCRIPTION
**Pull request by darkwing**
Fixes: #

Explanation:  

We switched to using `getBigNumber` in this commit:  https://github.com/MetaMask/metamask-extension/commit/24d7d1ae2add7422e7544d9a28b841d0374c5b0d ; unfortunately not all use cases are providing required parameters to guarantee expected values, so we should tidy that up.

### CODE REVIEW

File: ui/app/components/ui/token-input/token-input.component.js [hunk 1]
```diff
@@ -80,6 +80,8 @@ export default class TokenInput extends PureComponent {
 
     const multiplier = Math.pow(10, Number(decimals || 0))
     const hexValue = multiplyCurrencies(decimalValue || 0, multiplier, {
+      multiplicandBase: 16,
```
Comment (Gudahtt) [inline ui/app/components/ui/token-input/token-input.component.js:83]: The multiplicand appears to be a decimal value here, not a hex value (the variable `decimalValue`).

File: ui/app/helpers/utils/conversion-util.js [hunk 2]
```diff
@@ -162,6 +162,11 @@ const conversionUtil = (
   })
 
 const getBigNumber = (value, base) => {
+  if (base === undefined) {
+    console.warn('getBigNumber problem -- no base provided!', value, base)
+    throw new Error('No base provided to getBigNumber')
```
[the remark this answers is not in the document]
Comment (rekmarks) [reply ui/app/helpers/utils/conversion-util.js:167]: Ah, I see that we type check the base parameters in a lot of different places. We should probably create an `isValidBase(value: unknown): boolean` utility function, and use it for all the base type checks in this file.

File: ui/app/helpers/utils/conversion-util.js [hunk 3]
```diff
@@ -173,6 +178,16 @@ const getBigNumber = (value, base) => {
...
+
+  if (aBase === undefined || bBase === undefined) {
+    console.warn(
+      'addCurrencies problem -- no aBase or bBase provided!',
+      aBase,
+      bBase,
+    )
+    throw new Error('No aBase or bBase provided to addCurrencies')
+  }
+
```
Comment (rekmarks) [inline ui/app/helpers/utils/conversion-util.js:190]: See comment for `getBigNumber` above.

File: ui/app/helpers/utils/conversion-util.js [hunk 4]
```diff
@@ -183,6 +198,16 @@ const addCurrencies = (a, b, options = {}) => {
...
   const { aBase, bBase, ...conversionOptions } = options
+
+  if (aBase === undefined || bBase === undefined) {
+    console.warn(
+      'subtractCurrencies problem -- no aBase or bBase provided!',
+      aBase,
+      bBase,
+    )
+    throw new Error('No aBase or bBase provided to subtractCurrencies')
+  }
```
Comment (rekmarks) [inline ui/app/helpers/utils/conversion-util.js:209]: See comment for `gitBigNumber` above.

File: ui/app/helpers/utils/conversion-util.js [hunk 5]
```diff
@@ -194,6 +219,17 @@ const subtractCurrencies = (a, b, options = {}) => {
...
+  if (multiplicandBase === undefined || multiplierBase === undefined) {
+    console.warn(
+      'multiplyCurrencies problem -- no multiplicandBase or multiplierBase provided!',
+      multiplicandBase,
+      multiplierBase,
+    )
+    throw new Error(
+      'No multiplicandBase or multiplierBase provided to multiplyCurrencies',
+    )
+  }
```
Comment (rekmarks) [inline ui/app/helpers/utils/conversion-util.js:231]: See comment for `gitBigNumber` above.

File: ui/app/helpers/utils/token-util.js [hunk 6]
```diff
@@ -217,6 +217,10 @@ export function getTokenFiatAmount(
   const currentTokenToFiatRate = multiplyCurrencies(
     contractExchangeRate,
     conversionRate,
+    {
+      multiplicandBase: 10,
+      multiplierBase: 16,
+    },
```
Comment (rekmarks) [inline ui/app/helpers/utils/token-util.js:223]: I'm still trying to figure out `contractExchangeRate`, but the base of `conversionRate` is definitely `10`.

File: ui/app/helpers/utils/transactions.util.js [hunk 7]
```diff
@@ -155,6 +155,8 @@ export function sumHexes(...args) {
   const total = args.reduce((acc, base) => {
     return addCurrencies(acc, base, {
       toNumericBase: 'hex',
+      aBase: 10,
+      bBase: 10,
     })
```
Comment (rekmarks) [inline ui/app/helpers/utils/transactions.util.js:160]: This function is called `sumHexes`, and both bases should be `16`. Following other suggestions in this review, the second parameter, `base`, should be renamed to `hexAmount`.

File: ui/app/pages/send/send-content/send-amount-row/amount-max-button/amount-max-button.utils.js [hunk 8]
```diff
@@ -12,8 +12,11 @@ export function calcMaxAmount({ balance, gasTotal, sendToken, tokenBalance }) {
     ? multiplyCurrencies(tokenBalance, multiplier, {
         toNumericBase: 'hex',
         multiplicandBase: 16,
+        multiplierBase: 10,
       })
     : subtractCurrencies(addHexPrefix(balance), addHexPrefix(gasTotal), {
         toNumericBase: 'hex',
+        aBase: 10,
+        bBase: 10,
```
Comment (rekmarks) [inline ui/app/pages/send/send-content/send-amount-row/amount-max-button/amount-max-button.utils.js:20]: These should both be `16` (the parameters/operands are hexademical strings).

**Review by rekmarks — CHANGES_REQUESTED** [review]
This is great cleanup work! There are some errors to fix and improvements to make, but all but one should be easy to implement!

File: ui/app/helpers/utils/token-util.js (hunk 6 above)
Comment (rekmarks) [reply ui/app/helpers/utils/token-util.js:223]: Update: both should be `10`. The `contractExchangeRate` comes from this data source (via the TokenRatesController): https://api.coingecko.com/api/v3/simple/token_price/ethereum?contract_addresses=0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48&vs_currencies=eth

File: ui/app/helpers/utils/confirm-tx.util.js [hunk 9]
```diff
@@ -34,24 +34,30 @@ export function getHexGasTotal({ gasLimit, gasPrice }) {
       toNumericBase: 'hex',
       multiplicandBase: 16,
       multiplierBase: 16,
+      aBase: 16,
+      bBase: 16,
```
Comment (rekmarks) [inline ui/app/helpers/utils/confirm-tx.util.js:38]: These lines should be deleted:
```suggestion
```

**Review by rekmarks — CHANGES_REQUESTED** [review]

File: ui/app/helpers/utils/conversion-util.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (darkwing) [reply ui/app/helpers/utils/conversion-util.js]: I left in the console.warn's for the sake of review debugging, but yep, I'll have those removed.

File: ui/app/helpers/utils/conversion-util.js [hunk 10]
```diff
@@ -54,6 +54,11 @@ const baseChange = {
   BN: (n) => new BN(n.toString(16)),
 }
 
+// Utility function for checking base types
+const isValidBase = (base) => {
+  return base !== undefined
```
Comment (rekmarks) [inline ui/app/helpers/utils/conversion-util.js:59]: Let's tighten up this type check:
```suggestion
  return base === 10 || base === 16
```

Comment (rekmarks) [reply ui/app/helpers/utils/conversion-util.js:59]: The alternative would be this, but I don't know if we ever want anything other than 10 and 16. Use this if you're not sure:
```javascript
return Number.isInteger(base) && base > 1
```

File: ui/app/helpers/utils/confirm-tx.util.js [hunk 11]
```diff
@@ -39,19 +39,23 @@ export function getHexGasTotal({ gasLimit, gasPrice }) {
...
+      bBase: 10,
     })
   })
 }
 
 export function addFiat(...args) {
-  return args.reduce((acc, base) => {
-    return addCurrencies(acc, base, {
+  return args.reduce((acc, ethAmount) => {
+    return addCurrencies(acc, ethAmount, {
```
Comment (rekmarks) [inline ui/app/helpers/utils/confirm-tx.util.js:54]:
```suggestion
  return args.reduce((acc, fiatAmount) => {
    return addCurrencies(acc, fiatAmount, {
```

**Review by rekmarks — CHANGES_REQUESTED** [review]
Looking good! Just a few more things

**Review by rekmarks — CHANGES_REQUESTED** [review]
One last thing:

For all the `Must specify [base] and [base]` messages, let's make them `Must specify valid [base] and [base]`.

**Review by rekmarks — APPROVED** [review]
LGTM!

**Review by kumavis — APPROVED** [review]

**Pull request merged**